File: frontend/src/features/guidance/model/geofence.ts
/**
 * Client mirror of the canonical server geometry in
 * `backend/locations/domain/geofence.py`.
 *
 * This file MUST stay arithmetically identical to that module. It is pinned to
 * it by `contracts/fixtures/geofence-distance.json`, asserted from both
 * languages (FR-043a). No second distance formula, no second Earth radius, and
 * no third classification state may be introduced anywhere.
 */

export const EARTH_RADIUS_M = 6371008.8;

export type GeofenceStatus = "INSIDE_GEOFENCE" | "OUTSIDE_GEOFENCE";

/**
 * Two distances closer together than this are indistinguishable. Pinned to
 * `tolerance_m` in `contracts/fixtures/geofence-distance.json` and asserted by
 * `frontend/tests/contract/geofence-parity.test.ts`.
 */
export const DISTANCE_TOLERANCE_M = 0.001;

export type GeofenceCoordinates = {
  latitude: number;
  longitude: number;
};

const DEGREES_PER_HALF_TURN = 180;

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / DEGREES_PER_HALF_TURN;
}

export function haversineDistanceM(
  origin: GeofenceCoordinates,
  destination: GeofenceCoordinates,
): number {
  const lat1 = toRadians(origin.latitude);
  const lon1 = toRadians(origin.longitude);
  const lat2 = toRadians(destination.latitude);
  const lon2 = toRadians(destination.longitude);
  const deltaLat = lat2 - lat1;
  const deltaLon = lon2 - lon1;
  const value =
    Math.sin(deltaLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(deltaLon / 2) ** 2;
  return EARTH_RADIUS_M * 2 * Math.asin(Math.min(1, Math.sqrt(value)));
}

export function classifyGeofence(distanceM: number, radiusM: number): GeofenceStatus {
  if (!Number.isFinite(distanceM) || distanceM < 0) {
    throw new Error("distance_m");
  }
  if (!Number.isFinite(radiusM) || radiusM <= 0) {
    throw new Error("radius_m");
  }
  return distanceM <= radiusM ? "INSIDE_GEOFENCE" : "OUTSIDE_GEOFENCE";
}

File: backend/locations/domain/geofence.py
from __future__ import annotations

import math
from dataclasses import dataclass
from decimal import Decimal
from enum import StrEnum

from locations.domain.locations import Coordinates

EARTH_RADIUS_M = 6_371_008.8


class LocationValidationResult(StrEnum):
    INSIDE_GEOFENCE = "INSIDE_GEOFENCE"
    OUTSIDE_GEOFENCE = "OUTSIDE_GEOFENCE"


@dataclass(frozen=True, slots=True)
class ValidatedPosition:
    latitude: Decimal
    longitude: Decimal
    accuracy_m: Decimal

    def __post_init__(self) -> None:
        Coordinates(self.latitude, self.longitude)
        if not self.accuracy_m.is_finite() or self.accuracy_m < 0:
            raise ValueError("accuracy_m")

    @property
    def coordinates(self) -> Coordinates:
        return Coordinates(self.latitude, self.longitude)


def haversine_distance_m(origin: Coordinates, destination: Coordinates) -> float:
    lat1, lon1 = math.radians(float(origin.latitude)), math.radians(float(origin.longitude))
    lat2, lon2 = (
        math.radians(float(destination.latitude)),
        math.radians(float(destination.longitude)),
    )
    delta_lat = lat2 - lat1
    delta_lon = lon2 - lon1
    value = (
        math.sin(delta_lat / 2) ** 2
        + math.cos(lat1) * math.cos(lat2) * math.sin(delta_lon / 2) ** 2
    )
    return EARTH_RADIUS_M * 2 * math.asin(min(1.0, math.sqrt(value)))


def classify_geofence(distance_m: float, radius_m: Decimal) -> LocationValidationResult:
    if not math.isfinite(distance_m) or distance_m < 0:
        raise ValueError("distance_m")
    if not radius_m.is_finite() or radius_m <= 0:
        raise ValueError("radius_m")
    return (
        LocationValidationResult.INSIDE_GEOFENCE
        if distance_m <= float(radius_m)
        else LocationValidationResult.OUTSIDE_GEOFENCE
    )


def geofences_overlap(
    first: Coordinates, first_radius_m: Decimal, second: Coordinates, second_radius_m: Decimal
) -> bool:
    return haversine_distance_m(first, second) <= float(first_radius_m + second_radius_m)
